lib/shelf-metadata.js

```diff
@@ -0,0 +1,258 @@
+    /**
+     * Metadata object for Shelf files.
+     */
+    class ShelfMetadata {
+        /**
+         * @param {string} uri The URI for the file we want metadata for.
+         * @param {string} authToken The token used to authenticate access.
+         */
+        constructor(uri, authToken) {
+            this.uri = uri;
+            this.authToken = authToken;
+            this.data = "";
+        }
+
+        /**
+         * Gets all of a file's metadata.
+         *
+         * @return {Promise}
+         */
+        getAll() {
+            var options;
+
+            logger.debug("In ShelfMetadata.getAll()");
+
+            options = this.getOptions();
+
+            logger.info(`Getting metadata for ${options.url}`);
+            logger.debug(`GET request options: ${JSON.stringify(options, null, 4)}`);
+
+            return request.get(options).then((response) => {
+                logger.debug("In request.get().then()");
+
+                return utils.handleResponse(response);
+            });
+        }
+
+        /**
+         * Gets one metadata property from the file.
+         *
+         * @param {string} property The file's metadata property to get.
+         * @return {Promise}
+         */
+        getProperty(property) {
+            var options;
+
+            logger.debug("In ShelfMetadata.getProperty()");
+            logger.debug(`Property param: ${property}`);
+
+            if (!property) {
+                throw new Error("Must provide a property to get.");
+            }
+
+            options = this.getOptions(property);
+
+            logger.info(`Getting metadata for ${options.url}`);
+            logger.debug(`GET request options: ${JSON.stringify(options, null, 4)}`);
+
+            return request.get(options).then((response) => {
+                logger.debug("In request.get().then()");
+
+                return utils.handleResponse(response);
+            });
+        }
+
+        /**
+         * Updates the file's metadata. Only updates mutable items.
+         *
+         * @param {Object} data The data to add to the file's metadata.
+         * @return {Promise}
+         */
+        updateAll(data) {
+            var options;
+
+            logger.debug("In ShelfMetadata.updateAll()");
+            logger.debug(`Data param: ${data}`);
+
+            if (!data) {
+                throw new Error("Must provide data to update.");
+            }
+
+            options = this.getOptions();
+            options = this.setOptionsData(options, data);
+
+            logger.info(`Updating metadata for ${options.url}`);
+            logger.debug(`PUT request options: ${JSON.stringify(options, null, 4)}`);
+
+            return request.put(options).then((response) => {
+                logger.debug("In request.put().then()");
+
+                return utils.handleResponse(response);
+            });
+        }
+
+        /**
+         * Updates the file's metadata. Only updates mutable items.
```

Comment:
As we discussed, a put will also post if the entry doesn't exist. This should be mentioned here.